Add onClick Play Functions to React Storybook Buttons

stevenanthonyrevo opened the pull request:
PR Fixes:

- Added interactions to Buttons, Buttons with Styles, Button with Icon, Button Variants Showcases
- I've added few interactions like handle check all buttons and verify the click occurred once

Tamir198 reviewed (changes requested):
Please remove package lock first and uae pnpm 

Then i will check the pr

Tamir198 commented on src/components/UI/Button/button.story.tsx:
```diff
@@ -18,6 +22,17 @@ export const button: Story = {
     children: 'Click me!',
   },
+  play: async ({ canvasElement, args }) => {
+    const onClickMock = args.onClick as Mock;
```
You are not mocking anything, just telling ts to declare the onclick method as Mock. 
Please use mocking function for this

stevenanthonyrevo replied:
I see. I didn't realize the mocking function needed to be called. I tried to revise in the [commit](https://github.com/Deadlink-Hunter/Broken-Link-Website/pull/156/commits/0f1116092b0975bc6789d5a3e69fe354c349140b)

Tamir198 commented on src/components/UI/Button/button.story.tsx:
```diff
@@ -18,6 +22,17 @@ export const button: Story = {
+    await expect(button).toBeEnabled();
+    await userEvent.click(button);
+    await expect(onClickMock).toHaveBeenCalled();
+    await expect(onClickMock).toHaveBeenCalledTimes(1);
```
Same here, if we are checking that something was called 1 time no need to check generally that it was called

stevenanthonyrevo replied:
Good point. I cleaned the expect calls.

Tamir198 commented on package.json:
```diff
@@ -21,8 +21,10 @@
+    "storybook": "^9.1.10",
     "@mantine/core": "8.2.1",
     "@mantine/hooks": "8.2.1",
+    "@storybook/test": "^8.6.14",
```
Those should be dev dependencies

stevenanthonyrevo replied:
I moved the dependencies to dev via [commit](https://github.com/Deadlink-Hunter/Broken-Link-Website/pull/156/commits/184704aafc6b33b34dbbd4660977fe0f04d2d4c5)

stevenanthonyrevo commented on src/components/UI/Button/button.story.tsx:
```diff
@@ -30,6 +45,17 @@ export const buttonWithStyles: Story = {
+    onClickMock.mockClear(); 
+    const canvas = within(canvasElement);
+    const button = canvas.getByRole('button', { name: 'Button with custom style' });
+    await expect(button).toBeInTheDocument();
```
The new method handles the canvas element and args and button name as params. This is done in the new [commit](https://github.com/Deadlink-Hunter/Broken-Link-Website/pull/156/commits/0f1116092b0975bc6789d5a3e69fe354c349140b).

Tamir198 commented on src/components/UI/Button/button.story.tsx:
```diff
@@ -9,51 +10,87 @@ const meta: Meta<typeof Button> = {
   },
   tags: ['autodocs'],
+  args: {
+    onClick: fn(),
```
Why is this part needed? we already checking the click function in the test function no ?

stevenanthonyrevo replied:
The check requires a mock function to work. It does check the mock function but it's still required because there would be no object to check the call count.

stevenanthonyrevo commented on src/components/UI/Button/button.story.tsx:
```diff
@@ -9,51 +10,87 @@ const meta: Meta<typeof Button> = {
 type Story = StoryObj<typeof meta>;
 
+async function ButtonClick(canvasElement: HTMLElement, args: any, buttonName: string) {
+  const onClickMock = args.onClick as Mock;
```
I think you mean "onClick: fn().mockReturnValue('Success!'),"

Tamir198 commented on src/components/UI/Button/button.story.tsx:
```diff
@@ -9,51 +10,87 @@ const meta: Meta<typeof Button> = {
 
 export const buttonVariantsShowcase: Story = {
-  render: () => (
+  render: ({ onClick }) => (
```
I think that the test is failing because there is no mock function implemented on the button, 
try to call `fn.mock()`  as the callback for the onClick and see if this will fix the test. 

If not you can add a todo - mock the on click function

stevenanthonyrevo replied:
I tested actions and interactions in the storybook. I was able to get a different experience. 

First we have an actions tab.

<img width="1069" height="289" alt="Screenshot 2025-10-17 at 10 51 30 AM" src="https://github.com/user-attachments/assets/cb219fa1-ffe1-4379-ada7-4b37ac0a5209" />

 Next we have an interactions tab. 

<img width="858" height="217" alt="Screenshot 2025-10-17 at 11 10 44 AM" src="https://github.com/user-attachments/assets/4afe84ee-85d9-4676-a6ec-b58cf858a8e0" />

After clicking on the interaction we get the rest of the tests. 

<img width="901" height="265" alt="Screenshot 2025-10-17 at 11 10 50 AM" src="https://github.com/user-attachments/assets/96b16f72-e29b-4058-9fca-e49a638b097f" />

I had to remove the`fn` method because it can only have once source and I was able to get `argTypes` to work. 

I'm still new to Storybook, but I think is working now.

stevenanthonyrevo commented:
I cleaned up the code to utilize `fn` method directly. Removed some extra types such as `ButtonArgs` extended from `ButtonProps` and reviewed both action and interaction tabs working in storybook. 

Please let me know if there still missing tests. See [commit](https://github.com/Deadlink-Hunter/Broken-Link-Website/pull/156/commits/bcd50a2f11afd51e79b4b07f470e750bd5406a50).

stevenanthonyrevo commented on src/components/UI/Button/button.story.tsx:
```diff
@@ -9,15 +28,34 @@ const meta: Meta<typeof Button> = {
 export default meta;
 type Story = StoryObj<typeof meta>;
 
+async function ButtonClick(canvasElement: HTMLElement, buttonName: string) {
```
I've updated the name in the [commit](https://github.com/Deadlink-Hunter/Broken-Link-Website/pull/156/commits/b67a948a48561cd53ca7884e5874d80cf51bc4de).

The mock function is called via the updated named method `buttonClick` and now utilize the `Primary Button` as the child arg. This should work as expected. 

Passing the `onClick: fn()` doesn't return an action on the action tab. FYI.

stevenanthonyrevo commented:
@Tamir198 Thanks for the merge! It was great working with you. I'm continuing to work on Hacktoberfest, I'd be happy continue to work together.
